proof/refine/ARM/Finalise_R.thy

```diff
@@ -4362,62 +4489,388 @@ global_interpretation delete_one
+  unfolding update_sched_context_def
+  apply (rule corres_guard_imp)
+    apply (rule corres_symb_exec_l'[where Q'="\<lambda>r s. ko_at r scPtr s \<and> (\<exists>n. is_sc_obj n r)",
+                                    where P="\<lambda>s. obj_at \<top> scPtr s"])
+      apply (rule corres_guard_imp[where P'=R and Q'=R for R])
+        apply (rule_tac F="(\<exists>n. is_sc_obj n obj)" in corres_gen_asm)
+        apply (case_tac obj; simp add: is_sc_obj_def)
+        apply (rule setSchedContext_no_stack_update_corres[where f'=f'])
+           apply (clarsimp simp: x obj_at_def intro!: y z a)+
+     apply (fastforce simp: exs_valid_def get_object_def in_monad)
+    apply (wpsimp wp: get_object_wp)
+   apply (fastforce simp: obj_at_def)
+  apply simp
+  done
+
+lemma schedContextUnbindTCB_corres:
+  "corres dc (einvs and sc_tcb_sc_at bound sc_ptr)
+             (invs' and obj_at' (\<lambda>sc. bound (scTCB sc)) sc_ptr)
+          (sched_context_unbind_tcb sc_ptr) (schedContextUnbindTCB sc_ptr)"
+  apply (clarsimp simp: sched_context_unbind_tcb_def schedContextUnbindTCB_def
+                        sym_refs_asrt_def)
+  apply add_sym_refs
+  apply (rule corres_stateAssert_implied[where P'=\<top>, simplified])
+  apply (rule corres_guard_imp)
+     apply (rule corres_split_new[OF get_sc_corres])
+       apply (rule corres_assert_opt_assume_l)
+       apply (rule corres_assert_assume_r)
+       apply (prop_tac "scTCB sca = sc_tcb sc"; clarsimp)
+        apply (clarsimp simp: sc_relation_def)
+       apply (rule corres_split_new[OF gct_corres])
+         apply (rule corres_split_new[OF corres_when], clarsimp simp: sc_relation_def)
+            apply (rule rescheduleRequired_corres)
+           apply (rule corres_split_new[OF tcbSchedDequeue_corres])
+             apply (rule corres_split_new[OF tcb_release_remove_corres])
+               apply (rule corres_split_new[OF set_tcb_obj_ref_corresT])
+                    apply (clarsimp simp: tcb_relation_def)
+                   apply (rule ball_tcb_cap_casesI; simp)
+                  apply (clarsimp simp: tcb_cte_cases_def)
+                 apply (rule_tac sc'=sca in set_sc_obj_ref_no_stack_update_corres)
+                    apply (clarsimp simp: sc_relation_def objBits_def objBitsKO_def)+
+                apply wpsimp+
+       apply (case_tac sca; clarsimp)
+       apply (wpfix add: sched_context.sel)
+       apply simp
+      apply wpsimp+
+    apply (frule invs_valid_objs)
+    apply (auto simp: sc_at_pred_n_def obj_at_def is_obj_defs valid_obj_def
+                      valid_sched_context_def)[1]
+   apply normalise_obj_at'
+   apply (frule invs_valid_objs')
+   apply (auto simp: valid_obj'_def valid_sched_context'_def projectKOs
+              dest!: ko_at_valid_objs')
+  done
+
+lemma get_tcbSchedContext_corres:
+  "corres (=) (pspace_aligned and pspace_distinct and tcb_at t) \<top>
+          (get_tcb_obj_ref tcb_sched_context t) (threadGet tcbSchedContext t)"
```

Comment:
There is `threadget_corres` that should cover this case once you unfold `get_tcb_obj_ref`.
I would be happy not to have this lemma, but we could generalise it so that it’s not about `tcbSchedContext` specific and keep it (and move to KHeap_R).

Reply:
I think I'm happy to unfold it here.
It's probably not worth changing at the moment, but would it make things simpler if `get_tcb_obj_ref` was an abbreviation?

Reply:
I changed my mind and did
```
lemmas get_tcb_obj_ref_corres
  = threadget_corres[where 'a="obj_ref option", folded get_tcb_obj_ref_def]
  ```